rotary/rules/polish.smk

```diff
@@ -135,10 +150,26 @@ rule polish_polca:
         mem=config.get("memory")
     shell:
         """
+        printf "\n\n### Replace bwa with bwa-mem2 ###\n" >> {log}
```

Comment:
I just noticed that a standalone version of POLCA, re-implemented in Python, was released that is called `pypolca`: https://github.com/gbouras13/pypolca (a [conda install](https://bioconda.github.io/recipes/pypolca/README.html#package-pypolca) also seems to exist)
This is now the recommended tool for polishing after Polypolish, in place of POLCA: https://github.com/rrwick/Polypolish/wiki
It also lets you set the output directory, unlike the original POLCA.
Looks like this was made by the hybracter team as part of their hybrid assembly pipeline.

What would you think about using this in place of POLCA?
It still relies on bwa, but the code you've written here could be used to swap out bwa for bwa-mem2 (or we might be able to directly raise an issue with the tool's authors to add bwa-mem2 support).

Reply:
One additional thought: because POLCA/pypolca was built with bwa (not bwa-mem2), I wonder if its polishing results could change a little after swapping to bwa-mem2. Do you think it might be worthwhile to do some benchmarking before swapping to bwa-mem2 specifically for POLCA/pypolca? I think the other changes to bwa-mem2 you've made in this PR all look fine.

Reply:
> I just noticed that a standalone version of POLCA, re-implemented in Python, was released that is called `pypolca`: https://github.com/gbouras13/pypolca (a [conda install](https://bioconda.github.io/recipes/pypolca/README.html#package-pypolca) also seems to exist) This is now the recommended tool for polishing after Polypolish, in place of POLCA: https://github.com/rrwick/Polypolish/wiki It also lets you set the output directory, unlike the original POLCA. Looks like this was made by the hybracter team as part of their hybrid assembly pipeline.

Let's move over to `pypolca`. Looking at the original `polca`'s github repo it looks like it was developed using poor software development practices. I was going to ask you about a replacement for stability reasons. So, let's go with the reimplementation.

Reply:
> What would you think about using this in place of POLCA? It still relies on bwa, but the code you've written here could be used to swap out bwa for bwa-mem2 (or we might be able to directly raise an issue with the tool's authors to add bwa-mem2 support).

@jmtsuji I would keep my code in for now and then raise the issue with the devs simutaneously.

Reply:
> One additional thought: because POLCA/pypolca was built with bwa (not bwa-mem2), I wonder if its polishing results could change a little after swapping to bwa-mem2. Do you think it might be worthwhile to do some benchmarking before swapping to bwa-mem2 specifically for POLCA/pypolca? I think the other changes to bwa-mem2 you've made in this PR all look like they should have limited to no impact to pipeline results.

@jmtsuji You would get changes between vanilla bwa code versions as well. For example, last year they inserted a bug into BTW that slightly changed the read mapping results in a way that was not intended. Also this would affect poly-polish as well since it takes similar read maps for input.

> > I just noticed that a standalone version of POLCA, re-implemented in Python, was released that is called `pypolca`: https://github.com/gbouras13/pypolca (a [conda install](https://bioconda.github.io/recipes/pypolca/README.html#package-pypolca) also seems to exist) This is now the recommended tool for polishing after Polypolish, in place of POLCA: https://github.com/rrwick/Polypolish/wiki It also lets you set the output directory, unlike the original POLCA. Looks like this was made by the hybracter team as part of their hybrid assembly pipeline.
> 
> @jmtsuji Let's move over to `pypolca`. Looking at the original `polca`'s github repo it looks like it was developed using poor software development practices. I was going to ask you about a replacement for stability reasons. So, let's go with the reimplementation.

@jmtsuji Note that the two tools produce slightly different results: https://github.com/gbouras13/pypolca/blob/main/benchmarking.md

Reply:
> > One additional thought: because POLCA/pypolca was built with bwa (not bwa-mem2), I wonder if its polishing results could change a little after swapping to bwa-mem2. Do you think it might be worthwhile to do some benchmarking before swapping to bwa-mem2 specifically for POLCA/pypolca? I think the other changes to bwa-mem2 you've made in this PR all look like they should have limited to no impact to pipeline results.
> 
> @jmtsuji You would get changes between vanilla bwa code versions as well. For example, last year they inserted a bug into BTW that slightly changed the read mapping results in a way that was not intended. Also this would affect poly-polish as well since it takes similar read maps for input.
> 
> > > I just noticed that a standalone version of POLCA, re-implemented in Python, was released that is called `pypolca`: https://github.com/gbouras13/pypolca (a [conda install](https://bioconda.github.io/recipes/pypolca/README.html#package-pypolca) also seems to exist) This is now the recommended tool for polishing after Polypolish, in place of POLCA: https://github.com/rrwick/Polypolish/wiki It also lets you set the output directory, unlike the original POLCA. Looks like this was made by the hybracter team as part of their hybrid assembly pipeline.
> > 
> > 
> > @jmtsuji Let's move over to `pypolca`. Looking at the original `polca`'s github repo it looks like it was developed using poor software development practices. I was going to ask you about a replacement for stability reasons. So, let's go with the reimplementation.
> 
> @jmtsuji Note that the two tools produce slightly different results: https://github.com/gbouras13/pypolca/blob/main/benchmarking.md

I think the polka workflow goes: BWA --> Samtools --> freebayes --> some filtering and post analysis. The software isn't fundamentally reliant on BWA. Theoretically, if it was programmed differently, it could take SAM/BAM files as input rather than taking the reads directly taking the raw reads. If it were implemented that way, it would be similar to polypolish and we could probably use the same read mapping step between the two of them (I would suggest this to the `pyploca` dev).

Reply:
> I think the polka workflow goes: BWA --> Samtools --> freebayes --> some filtering and post analysis. The software isn't fundamentally reliant on BWA. Theoretically, if it was programmed differently, it could take SAM/BAM files as input rather than taking the reads directly taking the raw reads. If it were implemented that way, it would be similar to polypolish and we could probably use the same read mapping step between the two of them (I would suggest this to the `pyploca` dev).
>
> Do you think it might be worthwhile to do some benchmarking before swapping to bwa-mem2 specifically for POLCA/pypolca? 

It might be worth some benchmarking or asking the `pyploca` dev.

Reply:
@LeeBergstrand OK, let's switch over to `pypolca`! (I also was also thinking it would be nice to eventually switch away from using POLCA.) I will leave it up to you whether we should make the switch in this PR versus wait until a future PR.

That also makes sense to ask the `pypolca` devs about adding bwa-mem2 support or allowing read mapping to be separated from the rest of `pypolca`. The `pypolca` devs might be able to re-run their existing benchmarks you linked me to, which would save us time of trying to make similar benchmarks ourselves. Good point that even changing bwa versions (or other tool versions) can affect the results. I'll make an issue in the `pypolca` repo.

Also, I agree, let's keep the code you wrote for swapping bwa and bwa-mem2 in this PR. We can use this for the time being.

Reply:
> @LeeBergstrand OK, let's switch over to `pypolca`! (I also was also thinking it would be nice to eventually switch away from using POLCA.) I will leave it up to you whether we should make the switch in this PR versus wait until a future PR.
> 
> That also makes sense to ask the `pypolca` devs about adding bwa-mem2 support or allowing read mapping to be separated from the rest of `pypolca`. The `pypolca` devs might be able to re-run their existing benchmarks you linked me to, which would save us time of trying to make similar benchmarks ourselves. Good point that even changing bwa versions (or other tool versions) can affect the results. I'll make an issue in the `pypolca` repo.
> 
> Also, I agree, let's keep the code you wrote for swapping bwa and bwa-mem2 in this PR. We can use this for the time being.

@jmtsuji I sent you some additional comments in a Signal DM.